File: go/luhn/luhn.go
// Package luhn implements the exercism.io Luhn number solution.
package luhn

import (
	"strconv"
	"strings"
)

func findDigitToSum(digit int, digitPos int) int {
	if digitPos%2 == 1 {
		var doubledDigit = digit * 2
		if doubledDigit > 9 {
			return doubledDigit - 9
		}

		return doubledDigit
	}

	return digit
}

func sumDigits(number string) (sum int) {
	var forwardIdx int
	for i := len(number) - 1; i >= 0; i = i - 1 {
		var digit = int(number[i] - '0')

		sum += findDigitToSum(digit, forwardIdx)

		forwardIdx++
	}

	return sum
}

// Valid returns true if number is a valid luhn number, false otherwise
func Valid(number string) bool {
	var cleanNumber = strings.Replace(number, " ", "", -1)
	var _, err = strconv.ParseInt(cleanNumber, 10, 64)

	if len(cleanNumber) <= 1 || err != nil {
		return false
	}

	return sumDigits(cleanNumber)%10 == 0
}

File: typescript/luhn/luhn.ts
import { stringify } from "querystring";

export default class Luhn {
  private static hasInvalidCharacters(number: string): boolean {
    const numberWithoutSpaces = number.replace(/\ /g, "");
    const numberWithOnlyDigits = number.replace(/\D/g, "");

    return numberWithoutSpaces != numberWithOnlyDigits;
  }

  private static doubleAndNormaliseIfNeeded(digit: number): number {
    const doubledDigit = digit * 2;
    if (doubledDigit > 9) {
      return doubledDigit - 9;
    } else {
      return doubledDigit;
    }
  }

  public static valid(number: string): boolean {
    if (this.hasInvalidCharacters(number)) {
      return false;
    }

    const numberWithOnlyDigits = Array.from(
      number.replace(/\D/g, "")
    ).reverse();

    const numberWithDoubledDigits = numberWithOnlyDigits.map(
      (digit: string, i: number) => {
        const isEvenPlacedDigit = i % 2 == 0;
        if (isEvenPlacedDigit) {
          return Number.parseInt(digit);
        } else {
          return this.doubleAndNormaliseIfNeeded(Number.parseInt(digit));
        }
      }
    );

    const sumOfDigits = numberWithDoubledDigits.reduce(
      (total: number = 0, digit: number) => {
        return total + digit;
      }
    );

    return sumOfDigits >= 10 && sumOfDigits % 10 == 0;
  }
}
